runtime/registry.go

```diff
@@ -79,11 +72,148 @@ func (r *Runtime) CreateInstance(ctx context.Context, inst *drivers.Instance) er
 	return nil
 }
 
-func (r *Runtime) DeleteInstance(ctx context.Context, instanceID string) error {
-	err := r.Registry().DeleteInstance(ctx, instanceID)
+func (r *Runtime) DeleteInstance(ctx context.Context, instanceID string, dropDB bool) error {
+	inst, err := r.Registry().FindInstance(ctx, instanceID)
 	if err != nil {
 		return err
 	}
 
-	return nil
+	svc, err1 := r.catalogCache.get(ctx, r, instanceID)
+	var dropErrors []error
+	if dropDB {
+		// drop all ingested data
+		// should we insert complete data in one schema so it will be easy to drop the schema ?
+		// NOTE : Database file does not free space when tables are dropped issue #1099 in duckdb
+		entries := svc.Catalog.FindEntries(ctx, inst.ID, drivers.ObjectTypeUnspecified)
+		for _, entry := range entries {
+			if entry.Type == drivers.ObjectTypeTable {
+				table := entry.GetTable()
+				if table.Managed {
+					dropErrors = append(dropErrors, svc.Olap.Exec(ctx, &drivers.Statement{Query: fmt.Sprintf("DROP TABLE %s", table.Name)}))
+				}
+			}
+		}
+		// drop rill schema which deletes embedded catalog as well
+		dropErrors = append(dropErrors, svc.Olap.Exec(ctx, &drivers.Statement{Query: "DROP schema rill CASCADE"}))
+	}
+	// delete instance related data if catalog is not embedded
+	if !inst.EmbedCatalog {
+		dropErrors = append(dropErrors, svc.Catalog.DeleteInstanceEntries(ctx, instanceID))
+	}
+
+	// evict caches
+	r.evictCaches(ctx, inst)
+
+	// delete instance
+	err4 := r.Registry().DeleteInstance(ctx, instanceID)
+	return util.ReturnFirstErr(util.ReturnFirstErr(err1, err4), util.ReturnFirstErr(dropErrors...))
+}
+
+// EditInstance edits exisiting instance.
+// Confirming to put api specs, it is expected to send entire existing instance data.
+// The API compares and only evicts caches drivers or dsn is changed.
+// This is done to ensure that db handlers are not unnecessarily closed
+func (r *Runtime) EditInstance(ctx context.Context, inst *drivers.Instance) error {
+	olderInstance, err := r.Registry().FindInstance(ctx, inst.ID)
+	if err != nil {
+		return err
+	}
+
+	// 1. changes in olap driver or olap dsn
+	var olapConn drivers.Connection
+	olapChanged := olderInstance.OLAPDriver != inst.OLAPDriver || olderInstance.OLAPDSN != inst.OLAPDSN
+	if olapChanged {
+		// Check OLAP connection
+		olap, _, err := r.checkOlapConnection(inst)
+		if err != nil {
+			return err
+		}
+
+		// Prepare connections for use
+		err = olap.Migrate(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to prepare instance: %w", err)
+		}
+		olapConn = olap
+	}
+
+	// 2. embedCatalog disabled previously but enabled now
+	if inst.EmbedCatalog {
+		if olapConn == nil {
+			// getting exisiting connection
+			olapConn, err = r.connCache.get(ctx, inst.ID, inst.OLAPDriver, inst.OLAPDSN)
+			if err != nil {
+				return err
+			}
+		}
+		_, ok := olapConn.CatalogStore()
+		if !ok {
+			return errors.New("driver does not support embedded catalogs")
+		}
+	}
```

Comment:
Get rid of the `olapConn` dependency (pretty hard to reason about, and things are not being closed correctly), and just get using the cache – keep it simple